app/models/secret_storage.rb

```diff
@@ -49,39 +50,65 @@ def self.keys
     end
   end
 
-  require 'vault'
   class HashicorpVault
 
     VAULT_SECRET_BACKEND ='secret/'.freeze
-    # we don't really want other directories in here,
+    # we don't really want other directories in the key,
     # and there may be other chars that we find we don't like
     ENCODINGS = {"/": "%2F"}
 
+    # get and cache a copy of the client that has a token
+    def self.vault_client
+      @vault_client ||= VaultClient.new
+    end
+
     def self.read(key)
-      result = Vault.logical.read(vault_path(key))
+      result = vault_client.logical.read(VAULT_SECRET_BACKEND + key)
       raise(ActiveRecord::RecordNotFound) if result.data[:vault].nil?
       result = result.to_h
       result = result.merge(result.delete(:data))
       result[:value] = result.delete(:vault)
       result
     end
 
+    # and parse it
     def self.write(key, data)
-      Vault.logical.write(vault_path(key), vault: data[:value])
+      key = SecretStorage.parse_secret_key(key, :key)
+      vault_client.logical.write(vault_path(key, data[:environment_permalink], data[:deploy_group_permalink], data[:project_permalink]), vault: data[:value])
     end
 
     def self.delete(key)
-      Vault.logical.delete(vault_path(key))
+      vault_client.logical.delete(VAULT_SECRET_BACKEND + key)
     end
 
     def self.keys()
-      Vault.logical.list(VAULT_SECRET_BACKEND).map! { |key| convert_path(key, :decode) }
+      base_keys = vault_client.logical.list(VAULT_SECRET_BACKEND)
+      base_keys = keys_recursive(base_keys)
+      base_keys.map! { |secret_path| convert_path(secret_path, :decode) }
     end
 
     private
 
-    def self.vault_path(key)
-      VAULT_SECRET_BACKEND + convert_path(key, :encode)
+    def self.keys_recursive(keys)
+      until all_leaf_nodes?(keys)
+        keys.each do |key|
+          vault_client.logical.list(VAULT_SECRET_BACKEND + key).each do |new_key|
+            keys << key + new_key
+          end
+          # nuke the key if it's a dir and we have processed it.
+          keys.delete(key) if key[-1] == '/'
+        end
+      end
+      keys
+    end
```

Comment:
recursive would avoid checking leaf nodes multiple times and potentially looping forever if something goes wrong ...


```Ruby
def keys_recursive(keys)
  keys.flat_map do |key| 
    if k.end_with?('/') # a directory
      keys_recursive(vault_client.logical.list(VAULT_SECRET_BACKEND + key))
    else # a leaf ... we are done
      key
    end
end
```